### PR TITLE
fix: enforce BLS spec KeyValidate and normalize error handling

### PR DESCRIPTION
**Pull request by UdjinM6**
Per [draft-irtf-cfrg-bls-signature-05](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-05), KeyValidate (section 2.5) must reject the identity element and CoreVerify (section 2.7) must call KeyValidate before pairing.

- Reject identity public keys in CoreMPL::Verify and CoreMPL::AggregateVerify per KeyValidate (section 2.5), which CoreVerify (section 2.7) and CoreAggregateVerify (section 2.9) require before pairing; the branch also rejects identity signatures there as additional hardening.
- Add IsValid() and identity checks to CoreMPL::VerifySecure and LegacySchemeMPL::Verify/AggregateVerify for defense-in-depth, matching the hardened core verification path.
- Override VerifySecure in AugSchemeMPL to verify T-weighted individual public keys against per-signer augmented messages, since Aug signs H(pk_i || msg) per signer and collapsing to one combined public key would lose that augmentation.
- Make these byte-based verification paths return false on malformed input instead of throwing: CoreMPL::Verify, CoreMPL::AggregateVerify, BasicSchemeMPL::AggregateVerify(Bytes...), AugSchemeMPL::AggregateVerify(Bytes...), PopSchemeMPL::PopVerify(Bytes...), and PopSchemeMPL::FastAggregateVerify(Bytes...).
- Add identity-element rejection to PopSchemeMPL::PopVerify(G1Element, G2Element) in line with PopVerify (section 3.3.3) calling KeyValidate(PK), and move validation before hashing for early exit.
- Add per-public-key validity and identity checks to PopSchemeMPL::FastAggregateVerify; this enforces the non-identity public-key expectation implied by the PopVerify precondition behind section 3.3.4, even though FastAggregateVerify itself delegates to CoreVerify after aggregating.

## Summary by CodeRabbit

* **New Features**
  * Added a public secure aggregate verification overload for simplified single-call verification.

* **Bug Fixes**
  * Stronger input validation: default/identity keys and invalid signatures are rejected.
  * Byte parsing now fails gracefully (returns false) instead of throwing.
  * Hardened aggregation/verification flows with defensive pre-checks.

* **Tests**
  * Extensive tests for malformed inputs, identity-key rejection, and aggregation edge cases.

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
@coderabbitai review

**Comment by UdjinM6** [comment]
@coderabbitai review

**Comment by UdjinM6** [comment]
@coderabbitai review

**Comment by UdjinM6** [comment]
@coderabbitai review

**Review by PastaPastaPasta — DISMISSED** [review]
ACK: 509bf2fb9b69fb6f31722bdcca31f2eef39372a2

**Comment by UdjinM6** [comment]
Addressed https://github.com/dashpay/bls-signatures/pull/123#pullrequestreview-3932530545

@coderabbitai review

**Comment by UdjinM6** [comment]
>402-418: Minor: Signature is parsed twice.
>
>The signature is parsed at line 404, but at line 418 the raw signature Bytes is passed to CoreMPL::AggregateVerify, which will parse it again. This could be avoided by calling the element-based overload with signatureElement.

Should be fixed now.

@coderabbitai review

File: src/schemes.cpp [hunk 1]
```diff
@@ -621,6 +699,11 @@ bool PopSchemeMPL::FastAggregateVerify(const vector<G1Element>& pubkeys,
...
         return false;
     }
+    for (const auto& pk : pubkeys) {
+        if (!pk.IsValid() || pk == G1Element()) {
+            return false;
+        }
+    }
     // No VerifyAggregateSignatureArguments checks required here as we have exactly one pubkey and one message.
     return CoreMPL::Verify(CoreMPL::Aggregate(pubkeys), message, signature);
 }
```
Comment (thepastaclaw) [inline src/schemes.cpp:712]: **🟡 Suggestion: FastAggregateVerify omits signature identity/validity check at entry point**

Every other native-element verification path in this PR checks `!signature.IsValid() || signature == G2Element()` before doing work. `FastAggregateVerify(vector<G1Element>, Bytes, G2Element)` skips this and proceeds directly to `CoreMPL::Aggregate(pubkeys)` — an O(n) group addition — before delegating to `CoreMPL::Verify`, which does check the signature. Adding the check early avoids unnecessary aggregation on identity/invalid signatures and keeps the defense-in-depth pattern consistent across all entry points.

```suggestion
bool PopSchemeMPL::FastAggregateVerify(const vector<G1Element>& pubkeys,
                                       const Bytes& message,
                                       const G2Element& signature)
{
    if (pubkeys.size() == 0) {
        return false;
    }
    if (!signature.IsValid() || signature == G2Element()) {
        return false;
    }
    for (const auto& pk : pubkeys) {
        if (!pk.IsValid() || pk == G1Element()) {
            return false;
        }
    }
    return CoreMPL::Verify(CoreMPL::Aggregate(pubkeys), message, signature);
}
```

<sub>source: ['claude-general', 'claude-security']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/schemes.cpp`:
- [SUGGESTION] lines 695-709: FastAggregateVerify omits signature identity/validity check at entry point
  Every other native-element verification path in this PR checks `!signature.IsValid() || signature == G2Element()` before doing work. `FastAggregateVerify(vector<G1Element>, Bytes, G2Element)` skips this and proceeds directly to `CoreMPL::Aggregate(pubkeys)` — an O(n) group addition — before delegating to `CoreMPL::Verify`, which does check the signature. Adding the check early avoids unnecessary aggregation on identity/invalid signatures and keeps the defense-in-depth pattern consistent across all entry points.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

PR correctly implements BLS spec KeyValidate with consistent identity-element rejection across all verification paths. The AugSchemeMPL::VerifySecure per-signer augmented message fix is cryptographically sound. Three minor defense-in-depth gaps remain, none exploitable.

*Reviewed commit: 180a8cff*

🟡 2 suggestion(s) | 💬 1 nitpick(s)

_2 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/schemes.cpp`:
- [SUGGESTION] lines 695-709: FastAggregateVerify omits signature identity/validity check at entry point
  Every other native-element verification path in this PR checks `!signature.IsValid() || signature == G2Element()` before doing work. `FastAggregateVerify(vector<G1Element>, Bytes, G2Element)` skips this and proceeds directly to `CoreMPL::Aggregate(pubkeys)` — an O(n) group addition — before delegating to `CoreMPL::Verify`, which does check the signature. Adding the check early avoids unnecessary aggregation on identity/invalid signatures and keeps the defense-in-depth pattern consistent across all entry points.
- [SUGGESTION] lines 262-266: Uncaught FromBytes in CoreMPL::VerifySecure and AugSchemeMPL::VerifySecure inner loops
  Lines 264 and 630 call `G1Element::FromBytes()` on round-tripped serialized data from elements that already passed `IsValid()` and identity checks, so these should never throw in practice. However, every other `FromBytes` call in the PR's verification paths is wrapped in try-catch. Wrapping these two sites would complete the pattern and guard against hypothetical serialization round-trip bugs.
```
</details>

File: src/schemes.cpp (hunk 1 above)
Comment (UdjinM6) [reply src/schemes.cpp:712]: should be resolved via a78ad28ba7671bb34e62a03b3db9efd7eb227783

File: src/test.cpp [hunk 2]
```diff
@@ -1571,6 +1592,323 @@ TEST_CASE("CheckValid")
...
+    }
+
+    SECTION("Identity signature rejected")
+    {
+        auto sk = PopSchemeMPL().KeyGen(getRandomSeed());
+        auto pk = sk.GetG1Element();
+        auto msg = getRandomSeed();
+        vector<G1Element> pks = {pk};
+        REQUIRE(PopSchemeMPL().FastAggregateVerify(pks, Bytes(msg), G2Element()) == false);
+    }
```
Comment (thepastaclaw) [inline src/test.cpp:1909]: **💬 Nitpick: Test section placed under mismatched parent TEST_CASE**

The parent TEST_CASE at line 1881 is `"FastAggregateVerify rejects identity pubkeys in input list"` but this SECTION tests identity *signatures*. A reader scanning test names would not expect signature-identity coverage under a pubkey-identity heading. Consider renaming the TEST_CASE to something broader like `"FastAggregateVerify rejects identity elements"`.

<sub>source: ['claude-general']</sub>

Comment (thepastaclaw) [inline src/test.cpp:1909]: **🟡 Suggestion: No coverage for the !signature.IsValid() branch**

The test exercises `signature == G2Element()` but not `!signature.IsValid()`. The new guard at schemes.cpp:702 has two conditions — only the identity half is tested. A section constructing a malformed G2Element (e.g. via FromBytes with invalid data, caught and stored before IsValid is false) would exercise the other branch.

<sub>source: ['claude-general']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/test.cpp`:
- [SUGGESTION] lines 1902-1909: No coverage for the !signature.IsValid() branch
  The test exercises `signature == G2Element()` but not `!signature.IsValid()`. The new guard at schemes.cpp:702 has two conditions — only the identity half is tested. A section constructing a malformed G2Element (e.g. via FromBytes with invalid data, caught and stored before IsValid is false) would exercise the other branch.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Clean, correct incremental fix that adds early identity-signature and validity rejection to PopSchemeMPL::FastAggregateVerify, consistent with the pattern in CoreMPL::Verify (line 152) and CoreMPL::VerifySecure (line 246). No blocking issues. Two minor observations worth noting.

*Reviewed commit: a78ad28b*

🟡 1 suggestion(s) | 💬 1 nitpick(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/test.cpp`:
- [SUGGESTION] lines 1902-1909: No coverage for the !signature.IsValid() branch
  The test exercises `signature == G2Element()` but not `!signature.IsValid()`. The new guard at schemes.cpp:702 has two conditions — only the identity half is tested. A section constructing a malformed G2Element (e.g. via FromBytes with invalid data, caught and stored before IsValid is false) would exercise the other branch.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

Trivial incremental: renames test case to broader title and adds a new SECTION exercising the `!signature.IsValid()` branch of FastAggregateVerify, directly addressing both findings from the prior review. The malformed-G2Element construction pattern follows existing test conventions (lines 1579-1584, 1650-1654). No issues found.

*Reviewed commit: b2cfcdfe*

**Review by thepastaclaw — APPROVED** [review]
Incremental push addresses both prior findings (test case rename + malformed-signature branch coverage). LGTM.

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit b2cfcdfe)

File: src/schemes.cpp [hunk 3]
```diff
@@ -115,14 +115,20 @@ bool CoreMPL::Verify(const vector<uint8_t> &pubkey,
...
-    return CoreMPL::Verify(G1Element::FromBytes(pubkey), message, G2Element::FromBytes(signature));
+    G1Element pubkeyElement;
+    G2Element signatureElement;
+    try {
+        pubkeyElement = G1Element::FromBytes(pubkey);
+        signatureElement = G2Element::FromBytes(signature);
+    } catch (...) {
+        return false;
+    }
+    return CoreMPL::Verify(pubkeyElement, message, signatureElement);
```
Comment (thepastaclaw) [inline src/schemes.cpp:131]: **🟡 Suggestion: `catch (...)` in bytes overloads also swallows operational errors like missing BLS::Init()**

The new `try { ... } catch (...) { return false; }` blocks across `CoreMPL::Verify`, `CoreMPL::AggregateVerify`, `BasicSchemeMPL::AggregateVerify`, `AugSchemeMPL::AggregateVerify`, `PopSchemeMPL::PopVerify`, and `PopSchemeMPL::FastAggregateVerify` (bytes overloads) catch *every* exception, not just malformed-input ones. `G1Element::FromBytes`/`G2Element::FromBytes` call `BLS::CheckRelicErrors()`, which throws `std::runtime_error("Library not initialized properly. Call BLS::Init()")` (`src/bls.cpp:81-85`). After this PR those operational faults are indistinguishable from invalid signatures: callers see `false` and have no way to detect that the library was never initialized. Narrow the catch to the specific exception types thrown by element parsing (e.g. `std::invalid_argument`) and let `std::runtime_error`/other operational exceptions propagate so initialization bugs remain diagnosable.

<sub>source: ['codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/schemes.cpp`:
- [SUGGESTION] lines 121-131: `catch (...)` in bytes overloads also swallows operational errors like missing BLS::Init()
  The new `try { ... } catch (...) { return false; }` blocks across `CoreMPL::Verify`, `CoreMPL::AggregateVerify`, `BasicSchemeMPL::AggregateVerify`, `AugSchemeMPL::AggregateVerify`, `PopSchemeMPL::PopVerify`, and `PopSchemeMPL::FastAggregateVerify` (bytes overloads) catch *every* exception, not just malformed-input ones. `G1Element::FromBytes`/`G2Element::FromBytes` call `BLS::CheckRelicErrors()`, which throws `std::runtime_error("Library not initialized properly. Call BLS::Init()")` (`src/bls.cpp:81-85`). After this PR those operational faults are indistinguishable from invalid signatures: callers see `false` and have no way to detect that the library was never initialized. Narrow the catch to the specific exception types thrown by element parsing (e.g. `std::invalid_argument`) and let `std::runtime_error`/other operational exceptions propagate so initialization bugs remain diagnosable.
```
</details>

File: src/schemes.cpp [hunk 4]
```diff
@@ -232,6 +238,15 @@ bool CoreMPL::VerifySecure(const std::vector<G1Element>& vecPublicKeys,
...
                            const Bytes& message,
                            const bool fLegacy) {
+    for (const auto& pk : vecPublicKeys) {
+        if (!pk.IsValid() || pk == G1Element()) {
+            return false;
+        }
+    }
+    if (!signature.IsValid() || signature == G2Element()) {
+        return false;
+    }
```
Comment (thepastaclaw) [inline src/schemes.cpp:248]: **🟡 Suggestion: VerifySecure no longer accepts the empty aggregate that AggregateSecure still produces**

`CoreMPL::AggregateSecure({}, {}, msg)` returns `G2Element()` (identity) because `CoreMPL::Aggregate` over an empty vector yields identity. Previously `VerifySecure({}, identity, msg)` round-tripped to true through `AggregateVerify({identity_pk}, {msg}, identity_sig)` — `VerifyAggregateSignatureArguments` itself documents the empty-aggregate convention at `src/schemes.cpp:69-71` (zero pubkeys + zero messages + identity signature is GOOD). The new identity-signature rejection at lines 246-247 (and the same code path at `src/schemes.cpp:611-612` in `AugSchemeMPL::VerifySecure`) makes the same identity aggregate unverifiable, breaking the `AggregateSecure`/`VerifySecure` round-trip the public API previously supported. If empty secure aggregates are intentionally unsupported now, also reject `vecPublicKeys.empty()` in `AggregateSecure` so the contract is symmetric; otherwise the early identity-signature reject should allow the empty-pubkey case through.

<sub>source: ['codex']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/schemes.cpp`:
- [SUGGESTION] lines 237-248: VerifySecure no longer accepts the empty aggregate that AggregateSecure still produces
  `CoreMPL::AggregateSecure({}, {}, msg)` returns `G2Element()` (identity) because `CoreMPL::Aggregate` over an empty vector yields identity. Previously `VerifySecure({}, identity, msg)` round-tripped to true through `AggregateVerify({identity_pk}, {msg}, identity_sig)` — `VerifyAggregateSignatureArguments` itself documents the empty-aggregate convention at `src/schemes.cpp:69-71` (zero pubkeys + zero messages + identity signature is GOOD). The new identity-signature rejection at lines 246-247 (and the same code path at `src/schemes.cpp:611-612` in `AugSchemeMPL::VerifySecure`) makes the same identity aggregate unverifiable, breaking the `AggregateSecure`/`VerifySecure` round-trip the public API previously supported. If empty secure aggregates are intentionally unsupported now, also reject `vecPublicKeys.empty()` in `AggregateSecure` so the contract is symmetric; otherwise the early identity-signature reject should allow the empty-pubkey case through.
```
</details>

**Review by thepastaclaw — COMMENTED** [review]
## Code Review

PR substantively hardens verification across schemes by rejecting identity G1/G2 elements per the BLS KeyValidate/CoreVerify spec and normalizing malformed-byte handling to return false. Two legitimate concerns remain: the broad `catch (...)` in bytes overloads now hides operational errors like missing `BLS::Init()`, and the new identity-sig rejection in `VerifySecure` breaks the empty round-trip with `AggregateSecure`. Also worth a chain-state check before legacy identity-rejection lands in a Dash Core release.

*Reviewed commit: b2cfcdfe*

🟡 3 suggestion(s)

### 1 additional finding

<details>
<summary>🟡 <b>suggestion:</b> Legacy identity rejection is a consensus-relevant behavioral change — confirm via chain-state scan</summary>

`src/schemes.cpp (lines 750-767)`

Adding `pubkey == G1Element()` / `signature == G2Element()` rejection to `LegacySchemeMPL::Verify` and `LegacySchemeMPL::AggregateVerify` (line 781) is correct per the BLS spec, but `LegacySchemeMPL` exists specifically to keep historical Dash chain verification stable for pre-V19 data. Any historical artifact (masternode list entry, quorum commitment, IS/CL signature) that contains an identity public key or identity signature would previously verify and would now be rejected — causing a hard fork for nodes upgrading past this commit. In practice no rational signer produces identity elements and there is no known on-chain instance, but please confirm with a chain-state scan of observed BLS pubkeys/signatures before this lands in a Dash Core release. The Basic/Aug/Pop changes carry no equivalent risk because they post-date V19.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/schemes.cpp`:
- [SUGGESTION] lines 121-131: `catch (...)` in bytes overloads also swallows operational errors like missing BLS::Init()
  The new `try { ... } catch (...) { return false; }` blocks across `CoreMPL::Verify`, `CoreMPL::AggregateVerify`, `BasicSchemeMPL::AggregateVerify`, `AugSchemeMPL::AggregateVerify`, `PopSchemeMPL::PopVerify`, and `PopSchemeMPL::FastAggregateVerify` (bytes overloads) catch *every* exception, not just malformed-input ones. `G1Element::FromBytes`/`G2Element::FromBytes` call `BLS::CheckRelicErrors()`, which throws `std::runtime_error("Library not initialized properly. Call BLS::Init()")` (`src/bls.cpp:81-85`). After this PR those operational faults are indistinguishable from invalid signatures: callers see `false` and have no way to detect that the library was never initialized. Narrow the catch to the specific exception types thrown by element parsing (e.g. `std::invalid_argument`) and let `std::runtime_error`/other operational exceptions propagate so initialization bugs remain diagnosable.
- [SUGGESTION] lines 237-248: VerifySecure no longer accepts the empty aggregate that AggregateSecure still produces
  `CoreMPL::AggregateSecure({}, {}, msg)` returns `G2Element()` (identity) because `CoreMPL::Aggregate` over an empty vector yields identity. Previously `VerifySecure({}, identity, msg)` round-tripped to true through `AggregateVerify({identity_pk}, {msg}, identity_sig)` — `VerifyAggregateSignatureArguments` itself documents the empty-aggregate convention at `src/schemes.cpp:69-71` (zero pubkeys + zero messages + identity signature is GOOD). The new identity-signature rejection at lines 246-247 (and the same code path at `src/schemes.cpp:611-612` in `AugSchemeMPL::VerifySecure`) makes the same identity aggregate unverifiable, breaking the `AggregateSecure`/`VerifySecure` round-trip the public API previously supported. If empty secure aggregates are intentionally unsupported now, also reject `vecPublicKeys.empty()` in `AggregateSecure` so the contract is symmetric; otherwise the early identity-signature reject should allow the empty-pubkey case through.
- [SUGGESTION] lines 750-767: Legacy identity rejection is a consensus-relevant behavioral change — confirm via chain-state scan
  Adding `pubkey == G1Element()` / `signature == G2Element()` rejection to `LegacySchemeMPL::Verify` and `LegacySchemeMPL::AggregateVerify` (line 781) is correct per the BLS spec, but `LegacySchemeMPL` exists specifically to keep historical Dash chain verification stable for pre-V19 data. Any historical artifact (masternode list entry, quorum commitment, IS/CL signature) that contains an identity public key or identity signature would previously verify and would now be rejected — causing a hard fork for nodes upgrading past this commit. In practice no rational signer produces identity elements and there is no known on-chain instance, but please confirm with a chain-state scan of observed BLS pubkeys/signatures before this lands in a Dash Core release. The Basic/Aug/Pop changes carry no equivalent risk because they post-date V19.
```
</details>
